cpp/src/arrow/device.h

```diff
@@ -165,6 +224,8 @@ class ARROW_EXPORT MemoryManager : public std::enable_shared_from_this<MemoryMan
   static Result<std::shared_ptr<Buffer>> ViewBuffer(
       const std::shared_ptr<Buffer>& source, const std::shared_ptr<MemoryManager>& to);
 
+  virtual Result<std::shared_ptr<DeviceSync>> MakeDeviceSync(void* sync_event = nullptr);
```

Comment:
Nit: don't put default arguments on virtual functions. Also, if possible: for virtual functions which aren't pure virtual, make the default impl obvious by inlining
```suggestion
  virtual Result<std::shared_ptr<DeviceSync>> MakeDeviceSync() {
    return nullptr;
  }
  Result<std::shared_ptr<DeviceSync>> MakeDeviceSync(void *sync_event);
```